lib/src/main/java/io/ably/lib/types/BaseMessage.java

```diff
@@ -109,16 +109,17 @@ public void encode(ChannelOptions opts) throws AblyException {
 			data = Serialisation.gson.toJson((JsonElement)data);
 			encoding = ((encoding == null) ? "" : encoding + "/") + "json";
 		}
-		if(opts != null && opts.encrypted) {
-			if(data instanceof String) {
+		if(data instanceof String) {
+			if (opts != null && opts.encrypted) {
 				try { data = ((String)data).getBytes("UTF-8"); } catch(UnsupportedEncodingException e) {}
 				encoding = ((encoding == null) ? "" : encoding + "/") + "utf-8";
 			}
-			if(!(data instanceof byte[])) {
-				throw AblyException.fromErrorInfo(new ErrorInfo("Unable to encode message data (incompatible type)", 400, 40000));
-			}
+		} else if(!(data instanceof byte[])) {
+			throw AblyException.fromErrorInfo(new ErrorInfo("Unable to encode message data (incompatible type)", 400, 40000));
```

Comment:
Should this not be "Invalid message data or encoding", see https://github.com/ably/ably-common/blob/master/protocol/errors.json#L19